File: labs/2nd semestr/lab5/string.ts
export abstract class Figure {
    abstract area(): number;
    abstract perimeter(): number;
}

type Point = [number, number];

export class Trapezoid extends Figure {
    constructor(
        private A: Point,
        private B: Point,
        private C: Point,
        private D: Point
    ) {
        super();
    }

    private distance(p1: Point, p2: Point): number {
        return Math.sqrt(
            Math.pow(p2[0] - p1[0], 2) + Math.pow(p2[1] - p1[1], 2)
        );
    }

    area(): number {
        const h = Math.abs(this.A[1] - this.C[1]);
        const a = this.distance(this.A, this.B);
        const b = this.distance(this.C, this.D);
        return 0.5 * (a + b) * h;
    }

    perimeter(): number {
        return (
            this.distance(this.A, this.B) +
            this.distance(this.B, this.C) +
            this.distance(this.C, this.D) +
            this.distance(this.D, this.A)
        );
    }
}

export class Circle extends Figure {
    constructor(private radius: number) {
        super();
    }

    area(): number {
        return Math.PI * this.radius * this.radius;
    }

    perimeter(): number {
        return 2 * Math.PI * this.radius;
    }
}

File: labs/2nd semestr/lab5/string.cs
using System;

public abstract class Figure
{
    public abstract double Area();
    public abstract double Perimeter();
}

public class Trapezoid : Figure
{
    private (double x, double y) A, B, C, D;

    public Trapezoid((double, double) a, (double, double) b,
                     (double, double) c, (double, double) d)
    {
        A = a; B = b; C = c; D = d;
    }

    private double Distance((double, double) p1, (double, double) p2)
    {
        return Math.Sqrt(Math.Pow(p2.Item1 - p1.Item1, 2) + Math.Pow(p2.Item2 - p1.Item2, 2));
    }

    public override double Area()
    {
        double h = Math.Abs(A.Item2 - C.Item2);
        double a = Distance(A, B);
        double b = Distance(C, D);
        return 0.5 * (a + b) * h;
    }

    public override double Perimeter()
    {
        return Distance(A, B) + Distance(B, C) + Distance(C, D) + Distance(D, A);
    }
}

public class Circle : Figure
{
    private double radius;

    public Circle(double r)
    {
        radius = r;
    }

    public override double Area()
    {
        return Math.PI * radius * radius;
    }

    public override double Perimeter()
    {
        return 2 * Math.PI * radius;
    }
}
